py/server/deephaven/table.py

```diff
@@ -43,6 +45,43 @@
 _JReverseAsOfJoinRule = jpy.get_type("io.deephaven.api.ReverseAsOfJoinRule")
 _JTableOperations = jpy.get_type("io.deephaven.api.TableOperations")
 
+# Dynamic Query Scope
+_JQueryScope = jpy.get_type("io.deephaven.engine.table.lang.QueryScope")
+_JUnsynchronizedScriptSessionQueryScope = jpy.get_type(
+    "io.deephaven.engine.util.AbstractScriptSession$UnsynchronizedScriptSessionQueryScope")
+_JPythonScriptSession = jpy.get_type("io.deephaven.integrations.python.PythonDeephavenSession")
+_j_script_session = jpy.cast(_JQueryScope.getScope(), _JUnsynchronizedScriptSessionQueryScope).scriptSession()
+_j_py_script_session = jpy.cast(_j_script_session, _JPythonScriptSession)
+_deephaven_pkg_prefix = __file__
+
+
+@contextlib.contextmanager
+def _query_scope_ctx():
+    """A context manager to set/unset query scope based on the scope of the most immediate application code that invokes
+    Table operations."""
+
+    # locate the innermost Deephaven frame (i.e. any of the table operation methods that use this context manager)
+    outer_frames = inspect.getouterframes(inspect.currentframe())[1:]
+    for i, (frame, filename, *_) in enumerate(outer_frames):
+        if filename and filename == __file__:
+            break
+
+    # we deem any module/function/method which calls a wrapped formula/filter-evaluating Table op as an application
+    # frame. This holds true and makes sense even when such module/function/method is part of the Deephaven package, and
+    # should not be a hurdle for developing such code as function/method parameters are now part of the query scope
+    if len(outer_frames) > i + 2:
+        app_frame, *_ = outer_frames[i + 1]
+        scope_dict = app_frame.f_globals.copy()
+        scope_dict.update(app_frame.f_locals)
+        try:
+            _j_py_script_session.pushScope(scope_dict)
+            yield
+        finally:
+            _j_py_script_session.popScope()
+    else:
+        # in the __main__ module, use the default main global scope
+        yield
```

Comment:
1) Are function locals in the main global scope handled by the if-block?
2) Are we properly handling values in an `if __name__ == "__main__":` block?

I think we are ok on these, but I'm confirming, since I'm not 100% sure.  Are there any other weird cases to consider?

Reply:
1. yes
2. yes
3. AFAIK, no